File: src/nl/wernerkroneman/Drawy/Modelling/CompositeModel.java
package nl.wernerkroneman.Drawy.Modelling;

import nl.wernerkroneman.Drawy.ModelEditor.CommandResultSelector;

import java.util.HashSet;
import java.util.Set;

/**
 * The CompositeModel represents everything that DrawyMcDrawface
 * knows about the scene described so far.
 * 
 * It describes the scene on a very high level, and deals
 * in constraints rather than realisations of those constraints.
 */
public class CompositeModel extends Model {

    private Set<Component> components = new HashSet<>();
    private Set<Constraint> constraints = new HashSet<>();

    public CompositeModel(String name) {
        super(name);
    }

    public Component addComponentForModel(Model cube) {
        Component comp = new Component(cube);
        components.add(comp);
        return comp;
    }

    /**
     * Obtain a list of model instances in this model.
     *
     * @return A list of model instances.
     */
    public Set<Component> getComponents() {
        return components;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();

        for (Component comp : getComponents()) {
            builder.append(comp.toString());
        }

        return builder.toString();
    }

    public Set<Constraint> getConstraints() {
        return constraints;
    }

    public void addConstraint(RelativePositionConstraint relativePositionConstraint) {
        constraints.add(relativePositionConstraint);
    }

    public static class Component implements CommandResultSelector {
        private Model model;

        public Component(Model model) {
            this.model = model;
        }

        public Model getModel() {
            return model;
        }
    }
}

File: src/main/java/nl/wernerkroneman/Drawy/Modelling/CompositeModel.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy.Modelling

import java.util.*

/**
 * The CompositeModel represents everything that DrawyMcDrawface
 * knows about the scene described so far.
 *
 *
 * It describes the scene on a very high level, and deals
 * in constraints rather than realisations of those constraints.
 */
abstract class CompositeModel(id: UUID = UUID.randomUUID(),
                              name: String = "Anonymous composite") : Model(id, name) {

    abstract val components: MutableSet<Model>
}

class CompositeModelBase(id: UUID = UUID.randomUUID(),
                         name: String = "Anonymous composite",
                         override val components: MutableSet<Model> = HashSet<Model>()) : CompositeModel(id, name) {
    override fun derive(name: String): Model {
        return DerivedCompositeModel(UUID.randomUUID(), name, this)
    }
}

class DerivedCompositeModel(id: UUID = UUID.randomUUID(),
                            name: String = "Anonymous composite derivative",
                            val base: CompositeModel) : CompositeModel(id, name) {

    override fun derive(name: String): Model {
        return DerivedCompositeModel(UUID.randomUUID(), name, this)
    }

    override val components = MutableRelativeSet(base.components)
}